### DISCUSSION TITLE
Couldn't add trials mid-experiment

### DISCUSSION THREAD

**Discussion by steamededamame** [immersivecognition/unity-experiment-framework#158] [Q&A] [answered]
Hi, 

My experiment has a 'training' phase that only ends once the participant has done a few things correctly. I was trying to add more trials on the fly based on one of these conditions (i.e., "rescued_A" == false) but ran into a problem -- I was only able to add one extra trial to the block. When I tried to add another trial after that, I got this error message on Unity instead:  "KeyNotFoundException: The given key 'rescued_A' was not present in the dictionary."

In summary: 
Trial 1 - planned, results saved.
Trial 2 - added mid-experiment, results saved. 
Trial 3 - failed to run due to exception error.

The results for 'rescued_A' was logged properly in the results csv file so I can't figure out what went wrong as I copied your code directly from the Wiki: 

```c#
public class AddTrialsOnTheFly : MonoBehaviour
{
    public void AddMoreTrials(Trial trial)
    {    
        bool rescued_A = (bool)trial.result["rescued_A"];

        if (rescued_A == false)
        {
            Trial newTrial = trial.block.CreateTrial();

            newTrial.settings.SetValue("blockType", "bright");

            newTrial.Begin();
        }
        
        else
        {
        
            if (trial == trial.session.LastTrial)
            {
                trial.session.End();
            }
        else
            {
                trial.session.NextTrial.Begin();
            }

        }


    }
}

```

Please could you help me figure out what is wrong?

Thank you so much!

**Comment by jackbrookes** [comment]
Hi, could it be that your method `AddMoreTrials` is being called on the newly created trials? If so, the line

```cs
bool rescued_A = (bool)trial.result["rescued_A"];
```

Will fail because `rescued_A` does not exist. You could modify `AddMoreTrials` to something like this I suppose?

```cs
    public void AddMoreTrials(Trial trial)
    {    
        if (trial.number >= 1) return; // skip if beyond trial 1

        bool rescued_A = (bool)trial.result["rescued_A"];

        if (rescued_A == false)
        {
```

**Reply by steamededamame** [reply]
Hi again, 

Sorry - I didn't reply directly initially a couple of days ago so I'm reposting my reply:

I added the 
```c#  
if (trial.number >= 1) return; // skip if beyond trial 1
```
 line but it didn't work -- this time it went straight to the next (planned) trial even though the condition to add new trials was met. 

I have since made some changes to my code seeing I was unable to retrieve ``` bool rescued_A = (bool)trial.result["rescued_A"];``` as described earlier. This is the specification I'm working with, if it helps to understand my project: 


```c#
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using UXF;

public class ExperimentGenerator : MonoBehaviour
{
   public void GenerateExperiment(Session session)
    {
       Block firstBlock = session.CreateBlock(1); // the number of trials here could be higher - haven't decided. 
       // some settings

        Block secondBlock = session.CreateBlock(2); // This block is where I want to add trials on the fly.
       // some settings

       Block thirdBlock = session.CreateBlock(200); // created 200 trials to show something later in this reply
        // some settings 

    }
}
```

Instead of grabbing the results from ```trial.result["key"]```, I have created a coroutine that checks - right before each trial ends -  whether the conditions to add new trials to the ```secondBlock``` have been met. If the conditions are met, we'll add new trials.

```c#

    public IEnumerator AddNewTrialBool()
    {
        while (true)
        {
            yield return new WaitUntil(() => !uxf_ui.activeSelf && stressRatingGroup.activeSelf == true);

            rescued_A = rescueVictim_A.rescued_A; // example condition

            hasRun = false; // just another bool to prevent the 'AddNewTrials' code from running more than once.

            if (blockType == 2 && rescued_A == false)
            {
                newTrialConditionsMet = true;
            }
            else
            {
                newTrialConditionsMet = false;
            }
            yield return new WaitUntil(() => timer.inTrial == 0);
        }
    }
```

After that, I created another function that calls the ```AddMoreTrials(trial)``` coroutine only when we're in the second block, and this function is assigned to the 'On Trial End' event in the inspector: 

```c#
    public void AddMoreTrialsCoroutine(Trial trial)
    {
        if (addTrialCoroutineHasRun == false && blockType == 2)
        {
            addTrialCoroutineHasRun = true;
            StartCoroutine(AddMoreTrials(trial));          
        }
    }
```

... and the ```AddMoreTrials(trial)``` coroutine: 

```c#
    public IEnumerator AddMoreTrials(Trial trial)
    {      
        if (hasRun == false)
        {
            hasRun = true;
            if (newTrialConditionsMet == true && blockType == 2)
            {
                Trial newTrial = trial.block.CreateTrial();
                // some settings

               // I have removed 'newTrial.Begin()' because including it caused Unity to skip ahead, such that even though
              //       a new trial was successfully created, the player ended up playing a trial  in the next block. 
              // For some reason, 'trial.session.NextTrial.Begin();' also caused similar problems. 

                yield break;
            }
            else if (newTrialConditionsMet == false && blockType == 2)
            {        

                Block thirdBlock; 
                thirdBlock = session.GetBlock(3);
                thirdBlock.firstTrial.Begin();

                yield break;
            }
          
        }
        else
        {         
            yield break;
        }

    }

```


This works perfectly for the second block (bold in image below), but the trial-skipping problem is severe in the third block (highlighted rows) even though the  ```AddMoreTrials(trial)``` coroutine was no longer running: 

![image](https://github.com/immersivecognition/unity-experiment-framework/assets/122629664/cb66251f-552c-45ff-a4ce-808598cd9c15)

This was how many times the function that calls the coroutine was called in Unity, and it agrees with the results csv: 

![addMoreTrialsscreenshot](https://github.com/immersivecognition/unity-experiment-framework/assets/122629664/de62514e-dcf0-44a2-b54b-0f5d64d83084)

When I remove this from the 'On Trial End' event in the inspector, everything ran normally (albeit without the ability to add trials on the go) so it's most probably not my other scripts... maybe. 

Please do you know what's going on here? I've tried many things for past couple of days to no avail. 
I'm not very good at coding so will also need more hand-holding if you don't mind. 

Thank you for your time and sorry for such a long reply!

**Reply by jackbrookes** [reply]
The code seems overly complex for what you need to do, seems to me you should return to the simpler implementation previously. I don't see any need for the coroutines.
For example, in your event `On Trial End` you could simply check if the current block number is 2 and if so add more trials. 

```cs
if (trial.block.number != 2)
    return;

// check if we need to add another trial...
Trial newTrial = trial.block.CreateTrial();
```

Then after the event has run, I assume there's some condition or other event handler that starts the next trial.
Generally, I would advise you use a debugger in your IDE to step through the code to see what's going on!
https://docs.unity3d.com/Manual/ManagedCodeDebugging.html

**Reply by steamededamame** [reply, marked as answer]
Hi again, 

I still couldn't get it to work properly using the simpler implementation because it wouldn't retrieve ```trial.result```. I tried to find out if my other scripts were causing the problem but it was taking up too much of my time so I kinda gave up on that.

However, I have made it work using the coroutines and I also found out what caused the trial skipping problem after adding trials. Turns out that I forgot to remove ```BeginNextTrialSafe``` from the inspector. Since I also have it in my script, it's being called more than once, causing the trials to skip ahead.

Thank you again for your time :)
